### PR TITLE
Bootstrap command refactoring: improve debuggability (step 5)

### PR DESCRIPTION
**Pull request by Kobzol**
Continuation of https://github.com/rust-lang/rust/pull/127321.

This PR improves the debuggability of command execution, by improving the output logged when a command fails (it now includes the exact location where the command was created and where it was executed), and also by adding a "drop bomb", which will panic if a command was created, but not executed (which is probably a bug).

Here is how the output of a failed command looks like:
```
Command "git" "foo" "[bar]" (failure_mode=Exit, stdout_mode=Capture, stderr_mode=Capture) did not execute successfully.
Expected success, got exit status: 1
Created at: src/core/build_steps/compile.rs:1699:9
Executed at: src/core/build_steps/compile.rs:1699:58

STDOUT ----


STDERR ----
git: 'foo' is not a git command. See 'git --help'.
```

And this is what is printed if you forget to execute a command:
```
thread 'main' panicked at /projects/personal/rust/rust/src/tools/build_helper/src/drop_bomb/mod.rs:42:13:
command constructed at `src/core/build_steps/compile.rs:1699:9` was dropped without being executed: `git`
```

Best reviewed commit by commit.

Tracking issue: https://github.com/rust-lang/rust/issues/126819

r? @onur-ozkan

### CODE REVIEW

**Comment by Kobzol** [comment]
Ok, PR CI seems to have finally stabilized.

@rustbot ready

File: src/bootstrap/src/core/build_steps/dist.rs [hunk 1]
```diff
@@ -1060,11 +1060,7 @@ impl Step for PlainSourceTarball {
                 cmd.arg("--sync").arg(manifest_path);
             }
 
-            let config = if !builder.config.dry_run() {
-                cmd.capture().run(builder).stdout()
-            } else {
-                String::new()
-            };
+            let config = cmd.capture().run(builder).stdout();
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/dist.rs:1063]: this is pretty cool:)

File: src/bootstrap/src/core/build_steps/test.rs [hunk 2]
```diff
@@ -1352,6 +1352,53 @@ impl Step for CrateRunMakeSupport {
...
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(CrateBuildHelper { host: run.target });
+    }
+
+    /// Runs `cargo test` for build_helper.
+    fn run(self, builder: &Builder<'_>) {
+        let host = self.host;
+        let compiler = builder.compiler(builder.top_stage, host);
+
+        builder.ensure(compile::Std::new(compiler, host));
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/test.rs:1377]: Any reason for this?

File: src/bootstrap/src/utils/exec.rs [hunk 3]
```diff
@@ -142,19 +147,55 @@ impl BootstrapCommand {
     }
 
     /// Run the command, returning its output.
+    #[track_caller]
     pub fn run(&mut self, builder: &Build) -> CommandOutput {
         builder.run(self)
     }
+
+    /// Provides access to the stdlib Command inside.
+    /// All usages of this function should be eventually removed from bootstrap.
```
Comment (onur-ozkan) [inline src/bootstrap/src/utils/exec.rs:156]:
```suggestion
    /// FIXME: This function should be eventually removed from bootstrap.
```

File: src/bootstrap/src/core/build_steps/test.rs (hunk 2 above)
Comment (Kobzol) [reply src/bootstrap/src/core/build_steps/test.rs:1377]: You mean this specific line or the whole step? I copy-pasted the whole step from the `RunMakeSupport` step.

Comment (onur-ozkan) [reply src/bootstrap/src/core/build_steps/test.rs:1377]: This specific line. `RunMakeSupport` is using specific compiler in its type so it makes sense to ensure std for it. But we don't need to build the top stage compiler and std for testing build helper crate, using stage 0 compiler should be enough (and faster as we don't need to build entire tree for testing it). Same for `CrateRunMakeSupport` step too, it should use stage 0 compiler without needing to build compiler and std.

Comment (Kobzol) [reply src/bootstrap/src/core/build_steps/test.rs]: Ok, changed it to stage 0.

**Comment by Kobzol** [comment]
Rebased to fix conflicts.

**Comment by onur-ozkan** [comment]
Thanks!

@bors r+

**Comment by Kobzol** [comment]
Found two more cases of unused command executions.

@bors r=onur-ozkan rollup=never

**Pull request merged**